### ISSUE TITLE
Export fails for dates before 1980

### ISSUE THREAD

**Issue by stefan-sherwood** [GothenburgBitFactory/taskwarrior#3572]
#### To report a bug...

* What command(s) did you run?

`task add description:"Test date" modified:"1970-01-01T00:00:00Z"`

If I do the identical command with a slightly different date in the same format, it correctly adds the task with the given time/date,

```
$ task add description:"Test date" modified:"1970-01-01T00:00:01Z"
Created task 44.
```

Note that the date format I gave is correct, both according to [ISO-8601 rules](https://en.wikipedia.org/wiki/ISO_8601#Combined_date_and_time_representations) and according to [taskwarrior specification for supported dates](https://taskwarrior.org/docs/dates/#iso-8601-format-hahahugoshortcode12s3hbhb).

* What did you expect to happen?

Task to be added.

* What actually happened?

Error message: `'1970-01-01T00:00:00Z' is not a valid date in the 'Y-M-D' format.`

Task was not added.

* Paste the output of the ```task diag``` command.
```
task 3.0.2
   Platform: Linux

Compiler
    Version: 13.2.0
       Caps: +stdc +stdc_hosted +LP64 +c8 +i32 +l64 +vp64 +time_t64
 Compliance: C++17

Build Features
      CMake: 3.27.4
    libuuid: libuuid + uuid_unparse_lower
 Build type: release

Configuration
       File: /home/stefan/.taskrc (found), 1381 bytes, mode 100644
       Data: /home/stefan/.task (found), dir, mode 40755
    Locking: Enabled
         GC: Enabled
Hooks
     System: Enabled
   Location: /home/stefan/.task/hooks
             (-none-)

Tests
   Terminal: 350x113
       Dups: Scanned 51 tasks for duplicate UUIDs:
             No duplicates found
 Broken ref: Scanned 51 tasks for broken references:
             No broken references found
```

**Comment by djmitche** [comment]
I suspect this is because zero is treated as a sentinel value to indicate that the time wasn't converted, and the string ending in `:00Z` evaluates to zero as a UNIX timestamp.

**Comment by stefan-sherwood** [comment]
That makes sense. I am trying to run a script from the [syncall project](https://github.com/bergercookie/syncall) that synchronizes Google Keep with taskwarrior. It [uses a zero timestamp](https://github.com/bergercookie/syncall/blob/b9890d5c9e854c199c7efed4a691e156f5cfe604/syncall/filesystem/filesystem_file.py#L188) when there is no modification date, presumably to ensure that all other modification dates compare as being later.

If this is considered functioning as designed, I understand.

As an aside it's a bit strange that it interprets the format was being 'Y-M-D' when it pattern matches an ISO-8601 date/time format but that's just a nitpick about the error message.

**Comment by djmitche** [comment]
It's trying to interpret that using the `dateformat` config, but I think that allows some other, fixed formats as well (including all of the relative times like `tomorrow` or `soq` and apparently ISO-8601 times)

I think relying on `0` as a sentinel value was based on the idea that nobody was using Taskwarrior in 1970, but the truth is probably lost to time. Presumably we could make 0 valid and return an exception. But I don't know much about the "variant" and "evaluation" stuff in `src/columns/ColTypeDate.cpp` to know how to do that.

The error message could use some improvement, for sure. Maybe if that just said "Could not parse .. as a date after 1970" it would at least be accurate? WDYT?

**Comment by stefan-sherwood** [comment]
Looping in the author of syncall @bergercookie to the conversation since this is really about getting the two projects to agree on what makes sense in this context. Having a value that represents "the beginning of time" makes sense to me and having that value be 0 seems to be the right thing to do. Obviously I could change it to pass in 1 but that seems like a hacky workaround.

IMO taskwarrior clients should not be subject to the effects of project-internal sentinel values. I'm not at all familiar with the code but in general if a `Variant` needs to be able to have a type before its value is set then [an additional boolean](https://github.com/GothenburgBitFactory/taskwarrior/blob/70714e8ca06ac66962c527a86fc23addef76d35a/src/Variant.h#L109) (e.g. `bool _isset`) to indicate there's a value is merited. Otherwise, there could be an [additional type](https://github.com/GothenburgBitFactory/taskwarrior/blob/70714e8ca06ac66962c527a86fc23addef76d35a/src/Variant.h#L43) 
 (e.g. `type_none`) added that indicates it doesn't have a type.

In terms of the error message, my primary complaint is that it sounds like the reason it's not working is because it's comparing against a different format (YYYY-MM-DD) when ISO-8601 actually is accepted. The [comment on the code](https://github.com/GothenburgBitFactory/taskwarrior/blob/70714e8ca06ac66962c527a86fc23addef76d35a/src/columns/ColTypeDate.cpp#L189) implies the same thing. The simplest change to change the error message to:

`'1970-01-01T00:00:00Z' is not a valid date. See https://taskwarrior.org/docs/dates for more info.`

And also update the documentation to indicate what date ranges are accepted since ISO-8601 itself has no such restriction.

As another side note, the syncall project actually uses the now-deprecated [truncated ISO-8601 format](https://en.wikipedia.org/wiki/ISO_8601#Truncated_representations_(deprecated)). And taskwarrior accepts it, even though it's not listed in [the taskwarrior documentation](https://taskwarrior.org/docs/dates/#iso-8601-format-hahahugoshortcode11s3hbhb) as accepted.

**Comment by djmitche** [comment]
Hopefully someone more familiar with the Variant stuff can fix the internal sentinel values.

But the change to the error message and the two docs fixes sound like great PR(s)!

**Comment by stefan-sherwood** [comment]
I'm happy to do that but I cannot find where the documentation at https://taskwarrior.org/docs/dates/ is located to edit.

**Comment by djmitche** [comment]
It's in the https://github.com/GothenburgBitFactory/tw.org repo.

**Comment by tbabej** [comment]
@stefan-sherwood did we end up changing the docs here?

**Comment by tbabej** [comment]
The problem with using beginning of unix time as an actual value reproduced on 3.1.0.

**Comment by stefan-sherwood** [comment]
@tbabej I have not made the error/documentation change because I was hoping for an actual solution (aka supporting ISO-8601 dates as described in the doc). Putting a "btw, dates prior to 1970 are not supported" is a band-aid, not a proper solution.

**Comment by tbabej** [comment]
I thought the issue was about trying to use the 0 (=UTC midnight of 1970-01-01) as a value for the time stamp. Are you implying you have an use case for using dates before 1970?

**Comment by stefan-sherwood** [comment]
The documentation says that ISO-8601 dates are accepted but only a subset of them is actually accepted. Further, the error message for when you give a valid ISO-8601 date that is not accepted is confusing.

I ran across this problem because a separate software package ([syncall](https://github.com/bergercookie/syncall)) I am trying to use is using t=0. I haven't personally had a direct use for pre-1970 dates but I am using it to keep track of events that happened in the past so if any of those turned out to be pre-1970 then I would run into this problem in that context.

**Comment by stefan-sherwood** [comment]
After some further digging, it seems that the t=0 date is being returned from Google Keep (via gkeepapi) for items that have no modified date. I modified syncall code to munge the dates and was able to get the tasks imported into taskwarrior.

However, taskwarrior has another related bug wherein tasks with [dates before 1980](https://github.com/GothenburgBitFactory/taskwarrior/blob/244513fad7fe1a88a6480c64ac0e1d562f794983/src/columns/ColTypeDate.cpp#L194) can be created but several operations -- most notably for me, `export` -- fail.

Apparently, this is due to a [fix to a similar issue](https://github.com/GothenburgBitFactory/taskwarrior/pull/3651).

related bug: https://github.com/GothenburgBitFactory/taskwarrior/issues/1849

```
$ task status:pending +test export
[
'31536000' is not a valid date in the '' format.
```

It would be nice if the error message included which task it was complaining about since I have hundreds of tasks.

```
$ task 53 info

Name          Value
ID            53
Description   Test old modified dates
Status        Pending
Entered       2025-02-02 10:08:09 (41min)
Last modified 1970-12-31 19:00:00 (54.1y)
Tags          test
Virtual tags  PENDING READY TAGGED UNBLOCKED
UUID          a50a3676-b1f2-43b4-bdbd-4b5de40a8056
Urgency       0.8
[tag_test      x]

    tags    0.8 *    1 =    0.8
                         ------
                            0.8```

**Comment by djmitche** [comment]
The DateTime class has some interesting choices, but I think untangling them would be rather difficult -- and should probably be done comprehensively rather than bit-by-bit.

The export question may be answered in https://github.com/GothenburgBitFactory/taskwarrior/pull/3770 or with something related. Let's try to fix that.

**Comment by stefan-sherwood** [comment]
It seems there have been a handful of issues regarding incorrect or imperfectly synchronized guard rails to protect against the limitations of the date handling. Rather than add yet another guard rail I think a better solution would be for task warrior to be able to accept any dates. 

As a nice side-effect a lot of paranoid code could be removed and fixing the unclear/misleading error messages wouldn't be necessary.

**Comment by tbabej** [comment]
While the suggestion to support "any dates" might sound simple in its essence, I would argue the ability to support it comes with the tradeoffs whose costs might not warrant the added benefit.

Dates are currently internally stored as 64-bit timestamps. In order to store an arbitrary date, we'd have to use a numeric representation that would allow numbers of arbitrary sizes (commonly referred to as bignum, or arbitrary-precision arithmetic). That would come at a performance benefit  -- at a point where we have already slowed down Taskwarrior performance compared to the past, and an added implementation complexity.

Parsing the dates would also become more difficult, as soon as we move beyond 4-digit years, the parsers logic would have to be improved to not conflict with other types that might be represented by a sequence of digits.

I wholeheartidly agree about making the experience more usable from the user perspective and fixing the unclear error messages. But in essence I don't think dropping all guardrails is a reasonable thing to do, as they are there because of fundamental reasons of data types used to store the datetime task attributes, and not out of arbitrary choices.

**Comment by stefan-sherwood** [comment]
I don't think it's necessary to increase the storage size of dates. A 64-bit value used to store dates covers 292 billion years. When I said "any" dates, I really meant any [ISO-8601 dates](https://en.wikipedia.org/wiki/ISO_8601), which only cover 4-digit years (0000-9999).

This is undoubtedly already a solved problem elsewhere.

**Comment by djmitche** [comment]
I agree with @tbabej but for some more subtle reasons -- the command-line interface is full of heuristics and people have come to rely on them. The [1980 hack](https://github.com/GothenburgBitFactory/taskwarrior/blob/244513fad7fe1a88a6480c64ac0e1d562f794983/src/columns/ColTypeDate.cpp#L194) is just one example. And, the parsing is a little different everywhere -- whether that's parsing the import/export format, or parsing strings in the backend storage, or handling command-line input.

There probably is a way to do this systematically, but it would take a great deal of analysis to figure out how to do that, what corner cases it might break, and how to address that breakage.

I'm going to change the title of this issue to reflect a focus on the export issue from a few comments previous. If someone wants to take on a systematic approach to fixing date handling, that's great -- let's open a new issue.

**Comment by stefan-sherwood** [comment]
I certainly agree that we're talking about multiple issues here. But I don't understand the complication of supporting all [ISO-8601 dates](https://en.wikipedia.org/wiki/ISO_8601). 

There are two examples I know of: using t=0 as a sentinel (easily fixable in the manner I suggested, for example) and the date display parsing failing (per [#3651](https://github.com/GothenburgBitFactory/taskwarrior/pull/3651)) is there reason to believe there are other places in the code that rely on the limitation of dates?

I'd be happy to take a look and I'd like to understand a little better what the concern is.

**Comment by djmitche** [comment]
I ran into this bug myself: I accidentally typed `task x mod wait:31` when I meant `task x mod dep:31`. That didn't cause any problems until I noticed that bugwarrior wasn't working anymore, because when it tried to `task export` that task, `task export` failed.

In general, `task export` should not fail on any task, whether it contains `"wait":"1741147840"` or `"wait":"31"` or `"wait":"☃"`. And this appears to be fixed, I think from #3770:
```
⸩ t add foo wait:31
'31' is not a valid date.
⸩ t add foo wait:'1969-12-31 19:00:31'
'1969-12-31 19:00:31' is not a valid date.
⸩ t add foo wait:'1975-12-31 19:00:31'
'1975-12-31 19:00:31' is not a valid date.
```
Well, almost. I edited the DB to set `"wait":"0"`, and..
```
⸩ t 1 export
[
{"id":1,"description":"foo","entry":"20250305T041844Z","modified":"20250305T041844Z","status":"pending","uuid":"1107ccf5-2817-4b70-90dc-f20c06a6c4ae",,"urgency":0}
]
that double-comma is not OK.
